enclave_mgmt/data_collection.py

```diff
@@ -0,0 +1,316 @@
+
+    quiz_data = []
+    for course_id in grade_dict.keys():
+        for quiz in grade_dict[course_id]['quizzes']:
+            quiz_data.append({
+                'quiz_id': quiz['id'],
+                'quiz_name': quiz['name'],
+                'max_grade': quiz['sumgrades'],
+                'course_id': quiz['course']
+            })
+    return pd.DataFrame(quiz_data)
+
+
+def generate_attempts_summary_df(grade_dict):
+
+    attempt_data = []
+    for course_id in grade_dict.keys():
+        for _, users in grade_dict[course_id]['attempts'].items():
+            for _, quiz in users.items():
+                for attempt_summary in quiz['summaries']:
+                    attempt_data.append({
+                        'course_id': course_id,
+                        'user_id': attempt_summary['userid'],
+                        'quiz_id': attempt_summary['quiz'],
+                        'attempt_id': attempt_summary['id'],
+                        'attempt_number': attempt_summary['attempt'],
+                        'time_started': attempt_summary['timestart'],
+                        'time_finished': attempt_summary['timefinish'],
+                        'attempt_grade': attempt_summary['sumgrades']
+                    })
+    if len(attempt_data) != 0:
+        return pd.DataFrame(attempt_data)
+    else:
+        return pd.DataFrame(
+            columns=[
+                'course_id',
+                'user_id',
+                'quiz_id',
+                'attempt_id',
+                'attempt_number',
+                'time_started',
+                'time_finished',
+                'attempt_grade'
+            ]
+        )
+
+
+def generate_attempt_multichoice_response_df(grade_dict):
+
+    attempt_multichoice_response_data = []
+    for course_id in grade_dict.keys():
+        for _, users in grade_dict[course_id]['attempts'].items():
+            for _, quiz in users.items():
+                for _, attempt_detail in quiz['details'].items():
+                    for question in attempt_detail['questions']:
+                        for answer in question['answer']:
+                            attempt_multichoice_response_data.append(
+                                {
+                                    'course_id': course_id,
+                                    'user_id':
+                                        attempt_detail['attempt']['userid'],
+                                    'quiz_id':
+                                        attempt_detail['attempt']['quiz'],
+                                    'attempt_id':
+                                        attempt_detail['attempt']['id'],
+                                    'attempt_number':
+                                        attempt_detail['attempt']['attempt'],
+                                    'answer': answer,
+                                    'question_number': question['slot'],
+                                }
+                            )
```

Comment:
I do think for readability the nested for loops might be better. Based on discussions with Ripal in the past I don't think our data is large enough yet to warrant the swap to the list comprehension solution, but I do like knowing that it is available to us. We can chat with Ripal about it on Tuesday to see if he thinks the same. Thanks for the explanation and proposed solution!

Reply:
The resource consideration is a good thought. But, per both of your comments above, I think we don't really need to worry about any data size / memory usage issues today. Since we partition enclave data by academic year and the `.items()` here would be in the scope of a specific course JSON (and sparse), I'm not too worried about it in this code path / dataframe.